File: Algorithms/p120_Triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        n = len(triangle)
        table = [0]*n
        table_tmp = [0]*n
        table[0] = triangle[0][0]
        for i in range(1,n):
            table_tmp[0] = table[0]+triangle[i][0]
            for j in range(1,i):
                table_tmp[j] = min(table[j-1],table[j])+triangle[i][j]
            table_tmp[i] = table[i-1]+triangle[i][i]
            for j in range(i+1):
                table[j] = table_tmp[j]
        return min(table)

File: Algorithms/p120_Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();
        vector<int> table(n,0);
        vector<int> table_tmp(n);
        table[0] = triangle[0][0];
        for(int i=1;i<n;++i){
            table_tmp[0] = table[0]+triangle[i][0];
            for(int j=1;j<=i-1;++j){
                table_tmp[j] = min(table[j-1],table[j])+triangle[i][j];
            }
            table_tmp[i] = table[i-1]+triangle[i][i];
            for(int j=0;j<=i;++j){
                table[j] = table_tmp[j];
            }
        }
        return *min_element(table.begin(),table.end());
    }
};
